openshift/console
# Convert coFetch utilities to typescript and move to plugin SDK

**Fixes**: https://issues.redhat.com/browse/ODC-4550


**Analysis / Root cause**: We need to move utilities and components to plugin SDK so that plugins can start using them.


**Solution Description**: 
- Converted `coFetch` utilities to typescript and moved to plugin SDK.
- Added utilities for extending and creating custom error classes.
- Exposed `coFetch`, `coFetchJSON` and `coFetchText` utilities as public APIs.
- Re-exported `InternalReduxStore` from plugin SDK that can be used by monorepo packages. This can be exposed as internal API as well.
- Re-exported `coFetch` utilities from `@console/internal` to keep all the old imports intact.


**Screen shots / Gifs for design review**: No UI change.


**Unit test coverage report**: 


<img width="781" alt="Screenshot 2021-07-28 at 5 39 55 PM" src="https://user-images.githubusercontent.com/6041994/127319905-68c58f7d-3b37-48cc-8600-5622073ae74f.png">

**Test Setup**
- Build and run the dynamic demo plugin
- Check route `/test-utility-consumer`. The Page should render a JSON response from `/api/kubernetes/version`.
<img width="1792" alt="Screenshot 2021-08-06 at 12 30 17 AM" src="https://user-images.githubusercontent.com/6041994/128406754-239c4ce9-b7df-49e7-b506-402e3542f34b.png">


**Browser conformance**:
